### ISSUE TITLE
怎么指定2个cpu微调？

### ISSUE THREAD

**Issue by zhanghui-china** [InternLM/xtuner#349]
现在无论怎么改py样例文件，微调的时候只能用到1个gpu
如：
把 internlm2_chat_7b_qlora_oasst1_e3_copy.py 的 batch_size 为1的时候，执行：xtuner train  ./internlm2_chat_7b_qlora_oasst1_e3_copy.py --deepspeed deepspeed_zero2
![image](https://github.com/InternLM/xtuner/assets/63148804/e7dcf42d-c40b-474c-a502-c591d6ab4122)
![image](https://github.com/InternLM/xtuner/assets/63148804/787f1684-13d0-4729-a85e-8fcedeb880e8)
GPU 17G

batch_size 为2
![image](https://github.com/InternLM/xtuner/assets/63148804/f5008495-b73c-48f3-96db-907681c73ec9)
![image](https://github.com/InternLM/xtuner/assets/63148804/03e4eeb5-c8ec-47e1-8839-6ddb52e15e94)
GPU 20G

batch_size 为3
![image](https://github.com/InternLM/xtuner/assets/63148804/1e6fa513-5fb9-46ab-b5f4-6b92aa5c33c5)
![image](https://github.com/InternLM/xtuner/assets/63148804/4f33b3c3-6f16-402d-85c3-ad0b63cb2222)
GPU23G

虽然一个GPU接近占满，但是另一个GPU一直没用。

请问应该如何微调的时候让2个gpu显存都占满。

另外，batch_size改为2，3之后，貌似整体微调时间预估并没有明显变化。
请问如何可以大幅降低微调时间？

**Comment by zhanghui-china** [comment]
好的，多谢，我试一下。

**Comment by zhanghui-china** [comment]
> 1. 2个GPU训练，单机通常使用分布式训练（DIST）即可，即 `NPROC_PER_NODE=2 xtuner train XXXXX`
> 
> ```
> # On multiple GPUs
> (DIST) NPROC_PER_NODE=${GPU_NUM} xtuner train internlm2_chat_7b_qlora_oasst1_e3 --deepspeed deepspeed_zero2
> (SLURM) srun ${SRUN_ARGS} xtuner train internlm2_chat_7b_qlora_oasst1_e3 --launcher slurm --deepspeed deepspeed_zero2
> ```
> 
> 2. 增大 batchsize 后，速度并没有显著增加，可能是因为多条数据放到一个 mini-batch 后，存在很多 padding token 浪费计算资源。两种解决办法：
>    
>    1. 使用 `pack_to_max_length=True`，以避免 padding token。 https://github.com/InternLM/xtuner/blob/60fabeb2ba3fb552b8f2e2925353b328068434e5/xtuner/configs/internlm/internlm2_chat_7b/internlm2_chat_7b_qlora_oasst1_e3.py#L30
>    2. 使用 `xtuner.dataset.samplers.LengthGroupedSampler` 替换默认的 `DefaultSampler`，以此减少 padding token。即使用以下代码替换
>    
>    ```diff
>    -from mmengine.dataset import DefaultSampler
>    +from xtuner.dataset.samplers import LengthGroupedSampler
>    train_dataloader = dict(
>        batch_size=batch_size,
>        num_workers=dataloader_num_workers,
>        dataset=train_dataset,
>    -  sampler=dict(type=DefaultSampler, shuffle=True),
>    +  sampler=dict(
>    +       type=LengthGroupedSampler,
>    +       length_property='length',
>    +       per_device_batch_size=batch_size * accumulative_counts),
>        collate_fn=dict(type=default_collate_fn))
>    
>    train_dataloader = dict(
>        batch_size=batch_size,
>        num_workers=dataloader_num_workers,
>        dataset=llava_dataset,
>        
>        collate_fn=dict(type=default_collate_fn))
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    注意，上述两种方法选择任意一种即可，两种重叠使用不会带来任何益处

您好，pack_to_max_length=True 去不掉，去掉会报错：
<img width="968" alt="0fcaa14229404286105d3db5dc3066a" src="https://github.com/InternLM/xtuner/assets/63148804/8719be65-e480-4698-9c4e-927aecdbab26">

**Comment by zhanghui-china** [comment]
<img width="995" alt="1c9d957ebd36b169d84fc65b59b0466" src="https://github.com/InternLM/xtuner/assets/63148804/e8b38c7f-bf79-478d-8799-70b19fc40b06">

**Comment by LZHgrla** [comment]
@zhanghui-china  肯定不是删掉，不然会少变量，而是设为True 或 False

**Comment by zhanghui-china** [comment]
先实验修改 ack_to_max_length=False，并且 替换 xtuner.dataset.samplers.LengthGroupedSampler 之后
batch_size=4
<img width="986" alt="bb7b202fa907e220731c0ce27908ddd" src="https://github.com/InternLM/xtuner/assets/63148804/c383eeb2-be8c-4f02-8cff-4c3d233e39b2">
显存不足。

batch_size=3
显存23G
<img width="632" alt="9f3fa8fe755b8a6b24175bd9504b4dc" src="https://github.com/InternLM/xtuner/assets/63148804/61240fda-8324-42b3-b59b-5a356b52cb68">

<img width="997" alt="2234e564ca81c7a7feef82f9ae72481" src="https://github.com/InternLM/xtuner/assets/63148804/83f8cf1c-82fc-48d2-8bbd-04f2ea9800bf">
<img width="993" alt="e23529bebb44ffe526c086afa228acd" src="https://github.com/InternLM/xtuner/assets/63148804/f79765ce-a8de-4a65-a977-b56e6fdbf53e">

时间在递减，有点算不清。（原来是4小时多一点）
不知道是快了还是慢了。

**Comment by LZHgrla** [comment]
两种方法选择任意一种即可，两种重叠使用不会带来任何益处

要想最快使用 pack_to_max_length=True 就够了

**Comment by zhanghui-china** [comment]
True的时候：
batchsize=3
<img width="986" alt="051a19ef7b64f1619136299a21fd322" src="https://github.com/InternLM/xtuner/assets/63148804/2917c43f-2aec-432e-ae55-c2d0993d49e8">
<img width="632" alt="1650a99a131164cc8abc748c0484b2b" src="https://github.com/InternLM/xtuner/assets/63148804/f843af81-8ce6-45de-a763-5828dc729aa9">

**Comment by zhanghui-china** [comment]
为啥 前一种有33334，后一种才4040？差8倍之多？

**Comment by LZHgrla** [comment]
因为把多条数据样本做了pack

> 为啥 前一种有33334，后一种才4040？差8倍之多？

**Comment by zhanghui-china** [comment]
确实前面false那个，最后时间稳定在6小时多一点。比True的3种场景都要慢，True无论batch_size=1,2,3,时间都在4小时多一点。

**Comment by zhanghui-china** [comment]
<img width="972" alt="0ef3bc54902000fdd2917f50eeb85c5" src="https://github.com/InternLM/xtuner/assets/63148804/a637be1d-2911-4bf8-bd50-167f094c51cb">

**Comment by LZHgrla** [comment]
@zhanghui-china 
一个小建议，请尽量少张贴照片，转而使用 ``` ``` 来框住文本，不然会影响其他人阅读此issue～

比如
```
xxxxx
xxxxx
```
